beacon-chain/state/field_trie.go

```diff
@@ -0,0 +1,92 @@
+package state
+
+import (
+	"sync"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/state/stateutil"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/memorypool"
+)
+
+type FieldTrie struct {
+	*sync.Mutex
+	*reference
+	fieldLayers [][]*[32]byte
+	field       fieldIndex
+}
+
+func NewFieldTrie(field fieldIndex, elements [][]byte, length uint64) *FieldTrie {
+	return &FieldTrie{
+		fieldLayers: stateutil.ReturnTrieLayer(elements, length),
+		field:       field,
+		reference:   &reference{1},
+		Mutex:       new(sync.Mutex),
+	}
+}
+
+func (f *FieldTrie) RecomputeTrie(indices []uint64, elements [][]byte) ([32]byte, error) {
+	f.Lock()
+	defer f.Unlock()
+	var err error
+	var fieldRoot [32]byte
+	for _, idx := range indices {
+		root := bytesutil.ToBytes32(elements[idx])
+		f.fieldLayers[0][idx] = &root
+	}
+	fieldRoot, f.fieldLayers, err = stateutil.RecomputeFromLayer(f.fieldLayers, indices)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	return fieldRoot, nil
+}
+
+func (f *FieldTrie) RecomputeTrieVariable(indices []uint64, elements [][32]byte) ([32]byte, error) {
+	f.Lock()
+	defer f.Unlock()
+	var err error
+	var fieldRoot [32]byte
+
+	fieldRoot, f.fieldLayers, err = stateutil.RecomputeFromLayerVariable(elements, f.fieldLayers, indices)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	return fieldRoot, nil
+}
+
+func (f *FieldTrie) CopyTrie() *FieldTrie {
+	//f.Mutex.Lock()
+	//defer f.Mutex.Unlock()
+	if f.fieldLayers == nil {
+		return &FieldTrie{
+			field:     f.field,
+			reference: &reference{1},
+			Mutex:     new(sync.Mutex),
+		}
+	}
+	dstFieldTrie := [][]*[32]byte{}
+	switch f.field {
+	case randaoMixes:
+		dstFieldTrie = memorypool.GetTripleByteSliceRandaoMixes(len(f.fieldLayers))
+	case blockRoots:
+		dstFieldTrie = memorypool.GetTripleByteSliceBlockRoots(len(f.fieldLayers))
+	case stateRoots:
+		dstFieldTrie = memorypool.GetTripleByteSliceStateRoots(len(f.fieldLayers))
```

Comment:
this name is kinda confusing and hard to understand 'TripleByteSlice', maybe we can just rename to get StateRootsSlice ?